akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingGetStateSpec.scala

```diff
@@ -0,0 +1,182 @@
+    akka.cluster.sharding {
+      coordinator-failure-backoff = 3s
+      shard-failure-backoff = 3s
+      state-store-mode = "ddata"
+    }
+    """))
+
+  nodeConfig(first, second)(ConfigFactory.parseString(
+    """akka.cluster.roles=["shard"]"""))
+
+}
+
+class ClusterShardingGetStateSpecMultiJvmNode1 extends ClusterShardingGetStateSpec
+class ClusterShardingGetStateSpecMultiJvmNode2 extends ClusterShardingGetStateSpec
+class ClusterShardingGetStateSpecMultiJvmNode3 extends ClusterShardingGetStateSpec
+
+abstract class ClusterShardingGetStateSpec extends MultiNodeSpec(ClusterShardingGetStateSpecConfig) with STMultiNodeSpec with ImplicitSender {
+
+  import ClusterShardingGetStateSpec._
+  import ClusterShardingGetStateSpecConfig._
+
+  def initialParticipants = roles.size
+
+  def startShard(): ActorRef = {
+    ClusterSharding(system).start(
+      typeName = shardTypeName,
+      entityProps = Props(new ShardedActor),
+      settings = ClusterShardingSettings(system).withRole("shard"),
+      extractEntityId = extractEntityId,
+      extractShardId = extractShardId)
+  }
+
+  def startProxy(): ActorRef = {
+    ClusterSharding(system).startProxy(
+      typeName = shardTypeName,
+      role = Some("shard"),
+      extractEntityId = extractEntityId,
+      extractShardId = extractShardId)
+  }
+
+  def join(from: RoleName): Unit = {
+    runOn(from) {
+      Cluster(system).join(node(controller).address)
+    }
+    enterBarrier(from.name + "-joined")
+  }
+
+  "Inspecting cluster sharding state" must {
+
+    "join cluster" in {
+      join(controller)
+      join(first)
+      join(second)
+
+      // make sure all nodes has joined
+      awaitAssert {
+        Cluster(system).sendCurrentClusterState(testActor)
+        expectMsgType[CurrentClusterState].members.size shouldEqual 3
+      }
+
+      runOn(controller) {
+        startProxy()
+      }
+      runOn(first, second) {
+        startShard()
+      }
+
+      enterBarrier("sharding started")
+    }
+
+    "return empty state when no sharded actors has started" in {
+
+      awaitAssert {
+        val region = ClusterSharding(system).shardRegion(shardTypeName)
+        region ! ShardRegion.GetCurrentRegions
+        expectMsgType[ShardRegion.CurrentRegions].regions.size shouldEqual 0
+      }
+
+      enterBarrier("empty sharding")
+    }
+
+    "trigger sharded actors" in {
+      runOn(controller) {
+        val region = ClusterSharding(system).shardRegion(shardTypeName)
+
+        within(10.seconds) {
+          awaitAssert {
+            region ! Ping(1L)
+            expectMsg(Pong)
+            region ! Ping(2L)
+            expectMsg(Pong)
+            region ! Ping(3L)
+            expectMsg(Pong)
+            region ! Ping(4L)
+            expectMsg(Pong)
+          }
+        }
+      }
+
+      enterBarrier("sharded actors started")
+
+    }
+
+    "get shard state" in {
+      within(10.seconds) {
+        awaitAssert {
```

Comment:
It's often good to have a separate TestProbe for the things inside the awaitAssert, to avoid spilling over messages from one attempt to the next. Also note that the within defined here will currently be used for the things inside the awaitAssert, such as receiveWhile. That is often not what you intend, but with a separate TestProbe you will also get separate within timeouts for free.